src/decode-store.c

```diff
@@ -0,0 +1,320 @@
+    s->data_offset += data_size;
+    s->is_full = (s->data_offset == DECODE_STORE_SIZE - 1);
+}
+
+static void PrintInfo(struct DecodeStore *s)
+{
+    SCLogDebug("store %p used %u out of %u", s, s->data_offset, DECODE_STORE_SIZE);
+}
+
+static void Append(Packet *p, uint8_t *data, size_t data_size)
+{
+    SCLogDebug("start");
+    struct DecodeStore *s = GetActiveStore(&p->decode_store);
+    SCLogDebug("s %p", s);
+    if (!s) {
+        s = GetNewStore();
+        if (!s) {
+            // SET EVENT?
+            return;
+        }
+    }
+    SCLogDebug("s %p", s);
+    if (data_size + s->data_offset < DECODE_STORE_SIZE) {
+        Copy(s, data, data_size);
+        SCLogDebug("data fits, done");
+        PrintInfo(s);
+        return;
+    }
+    SCLogDebug("data does not fit, get additional store");
+
+    s->data[s->data_offset] = DECODE_STORE_TYPE_NEXT;
+    s->is_full = true;
+
+    s->next = GetNewStore();
+    if (!s->next) {
+        // SET EVENT?
+        return;
+    }
+    s = s->next;
+    Copy(s, data, data_size);
+    SCLogDebug("got additional store. Done");
+    PrintInfo(s);
+}
+
+
+typedef union DecodeStoreBuiltinVLAN {
+    struct {
+        uint16_t type:4;
+        uint16_t value:12;
+    };
+    uint8_t x[2];
+} DecodeStoreBuiltinVLAN;
+
+typedef union DecodeStoreBuiltinMPLS {
+    struct {
+        uint32_t type:4;
+        uint32_t pad:4;
+        uint32_t value:24;
+    };
+    uint8_t x[4];
+} DecodeStoreBuiltinMPLS;
+
+void DecodeStoreAddVLAN(Packet *p, const uint16_t vlan_id)
+{
+    DecodeStoreBuiltinVLAN v = {
+        .type = DECODE_STORE_TYPE_VLAN, .value = vlan_id
+    };
+    Append(p, v.x, sizeof(v.x));
+    SCLogDebug("id %u added (v.type %u v.value %u)", vlan_id, v.type, v.value);
+}
+
+void DecodeStoreAddMPLS(Packet *p, const uint32_t label)
+{
+    DecodeStoreBuiltinMPLS v = {
+        .type = DECODE_STORE_TYPE_MPLS, .value = label
+    };
+    Append(p, v.x, sizeof(v.x));
+    SCLogDebug("label %u added (v.type %u v.value %u)", label, v.type, v.value);
+}
+
+typedef struct DecodeStoreBuiltinTLV {
+    uint8_t type:4;
+    uint8_t pad:4;
+    uint8_t etype;
+    uint8_t len;
+} DecodeStoreBuiltinTLV;
+
+int DecodeStoreAddTLV(Packet *p, uint8_t etype, const uint8_t *data, uint8_t data_len)
+{
+    if (data_len + sizeof(DecodeStoreBuiltinTLV) >= (size_t)DECODE_STORE_SIZE)
+        return -EINVAL;
+
+    const size_t buf_size = sizeof(DecodeStoreBuiltinTLV) + data_len;
+    uint8_t buf[buf_size];
+    DecodeStoreBuiltinTLV hdr = {
+        .type = DECODE_STORE_TYPE_TLV, .pad = 0, .etype = etype, .len = data_len + 3
+    };
+    memcpy(buf, &hdr, sizeof(hdr));
+    memcpy(buf+sizeof(hdr), data, data_len);
+    Append(p, buf, buf_size);
+    SCLogDebug("TLV added %u datalen %u total %u", etype, data_len, hdr.len);
+    return 0;
+}
+
+static inline int GetTypeAndSize(const uint8_t *data, uint8_t size, uint8_t *type, uint8_t *out_size)
+{
+    assert(size >= 1);
+
+    *type = *data & 0x0f;
```

Comment:
Will 16 enough for types ?

Reply:
the idea was that there are some built-in types and then a flexible general purpose TLV type that allows 256 subtypes. The builtins should be extra compact/efficient.